### PR TITLE
Maintenance + NOT expression

### LINKED ISSUE #33: NOT operator
- To negate result of other expressions

#### Examples
- ```["NOT", ["IN", 1, [1, 2]]]``` -> **false**

### PR DESCRIPTION
**Pull request by davidhorak**
# Chnages
- Simplification of the codebase.
- Add Not logical expression.
- Breaking change: removed parser strict mode.
- Expressions are now responsible for validation their expected operands, rather than guarded in the parser, which should add us more flexibility.
- Add more tests simulation library usage.

#### References
Closes #33

### CODE REVIEW

**Comment by rajeevnaikte** [comment]
Can it be refactored to avoid committing `.d.ts` files?

**Comment by davidhorak** [comment]
> Can it be refactored to avoid committing `.d.ts` files?

Well the type files are part of the changes ... technically they could be rebuilt on master, during the publish time ... but they are changed in this PR, as the type files reflect the ts files

**Comment by rajeevnaikte** [comment]
Maybe add e2e test in https://github.com/briza-insurance/illogical/blob/master/src/__test__/unit/index.test.ts
for each type of operator.

**Comment by rajeevnaikte** [comment]
> > Can it be refactored to avoid committing `.d.ts` files?
> 
> Well the type files are part of the changes ... technically they could be rebuilt on master, during the publish time ... but they are changed in this PR, as the type files reflect the ts files

There is `"prepublishOnly": "npm run test && npm run build"`

**Comment by davidhorak** [comment]
> Maybe add e2e test in https://github.com/briza-insurance/illogical/blob/master/src/__test__/unit/index.test.ts
> for each type of operator.

How that would differ from this one https://github.com/briza-insurance/illogical/blob/master/src/parser/__test__/unit/index.test.ts ?

**Comment by davidhorak** [comment]
> > > Can it be refactored to avoid committing `.d.ts` files?
> > 
> > 
> > Well the type files are part of the changes ... technically they could be rebuilt on master, during the publish time ... but they are changed in this PR, as the type files reflect the ts files
> 
> There is `"prepublishOnly": "npm run test && npm run build"`

Yes ... but still "technically they could be rebuilt on master, during the publish time ... but they are changed in this PR, as the type files reflect the ts files"

**Comment by rajeevnaikte** [comment]
> > Maybe add e2e test in https://github.com/briza-insurance/illogical/blob/master/src/__test__/unit/index.test.ts
> > for each type of operator.
> 
> How that would differ from this one https://github.com/briza-insurance/illogical/blob/master/src/parser/__test__/unit/index.test.ts ?

Testing the way it will be used by consumers. I.e. creating an instance of `IllogicalEngine` and expected output to be `true` or `false`

File: src/__test__/unit/index.test.ts [hunk 1]
```diff
@@ -88,11 +89,87 @@ describe('Condition Engine', () => {
...
+        '$Name'
+      ],
+      inputs: [{
+        data: {
+          Name: undefined
+        },
+        expected: true
+      }, {
+        data: {
+          Name: 'David'
```
Comment (davidhorak) [inline src/__test__/unit/index.test.ts:156]: 😄 ... it looks like I've set it up for myself

Comment (rajeevnaikte) [reply src/__test__/unit/index.test.ts:156]: 🤣

**Pull request merged**